### PR TITLE
feat: extend API with ptr/len-string interfaces

### PR DESCRIPTION
**Pull request by supervacuus**
Fixes https://github.com/getsentry/sentry-native/issues/819.

### CODE REVIEW

**Comment by supervacuus** [comment]
Hi @Swatinem, can you give this a first once-over when you have time? (it is not urgent because I am unsure if I will have time to continue working on the PR before the easter holidays).

What the PR should do is rather apparent, but I was experimenting with how to de-duplicate the two paths in the code base. Sometimes the most trivial approach is to stay with two separate implementations (if the line count is low). Sometimes, it makes sense to express one in terms of the other. Sometimes I felt it would make the most sense to create a macro where parameters and parameter-dependent function calls are parameterized. 

When implementing one in terms of the other, I typically wrote the c-string interface in terms of the byte-string one because you only pay for an additional NULL-check and `strlen()`, which you'd have to do anyway later on if you don't end up in an error branch, vs. potentially paying double for allocation.

Could you tell me whether the macros make sense and whether you have different implementation favorites for those instances? The PR is incomplete; I want to add even more unit tests (some of the previous functions were "covered" but not tested, revealing a few places that led to NULL-induced access violations). I also haven't yet touched the windows-specific interfaces.

Thx!

**Comment by supervacuus** [comment]
The remaining macOS CI build also fails on `master`: https://github.com/getsentry/sentry-native/actions/runs/4542271760/jobs/8420339196; last week, the [macOS 11 image was updated](https://github.com/actions/runner-images/releases/tag/macOS-11%2F20230416.1). Of course, that doesn't mean there is no leak in the code, just that a change in this PR didn't introduce it.

I could not repro the leak using `leaks` on macOS, and the LLVM ASAN for aarch64 has some issues. But I will investigate further. Otherwise, I am open to any review suggestions :-)

**Comment by supervacuus** [comment]
haha, what? https://github.com/getsentry/sentry-native/actions/runs/4742464992/jobs/8420773714?pr=827 🤦

File: Makefile [hunk 1]
```diff
@@ -25,8 +25,10 @@ test-unit: update-test-discovery CMakeLists.txt
 	./unit-build/sentry_test_unit
 .PHONY: test-unit
 
+# SYSTEM_VERSION_COMPAT=0 is required (starting with BigSur) to get the correct macOS version for context assertions:
+# https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11_0_1-release-notes#Third-Party-Apps
 test-integration: setup-venv
-	.venv/bin/pytest tests --verbose
+	SYSTEM_VERSION_COMPAT=0 .venv/bin/pytest tests --verbose
```
Comment (supervacuus) [inline Makefile:31]: It is fine to leave this up to the environment and not add it to the `Makefile`. I added it for convenience when testing locally.

File: src/transports/sentry_transport_curl.c [hunk 2]
```diff
@@ -247,6 +247,7 @@ sentry__curl_send_task(void *_envelope, void *_state)
         }
     }
 
+    curl_easy_reset(curl);
```
Comment (markushi) [inline src/transports/sentry_transport_curl.c:250]: Any specific reason you added this?

**Review by markushi — APPROVED** [review]
Nice one, please have a look at my comments!

File: src/sentry_string.h [hunk 3]
```diff
@@ -124,6 +124,12 @@ sentry__string_clonen(const char *str, size_t n)
     return rv;
 }
 
+static inline char *
+sentry__string_clonen_or_null(const char *str, size_t n)
```
[the remark this answers is not in the document]
Comment (supervacuus) [reply src/sentry_string.h:128]: No particular reason; this naming was from before I entered the scene. The "n" is typically not separated by underscores in C libraries. But there is no reason not to make this consistent with the convention now introduced in the public API (given that this function is private). I am also not 100% happy with the naming asymmetry between the null-checked and inner functions. So... gonna fix this.

File: src/sentry_utils.c [hunk 4]
```diff
@@ -227,7 +226,7 @@ sentry__dsn_new(const char *raw_dsn)
     memset(dsn, 0, sizeof(sentry_dsn_t));
     dsn->refcount = 1;
 
-    dsn->raw = sentry__string_clone(raw_dsn);
+    dsn->raw = sentry__string_clonen_or_null(raw_dsn, raw_dsn_len);
     if (!dsn->raw || !dsn->raw[0] || sentry__url_parse(&url, dsn->raw) != 0) {
         goto exit;
```
[the remark this answers is not in the document]
Comment (supervacuus) [reply src/sentry_utils.c:231]: Most `DSN` fields must be deallocated (because `sentry__url_parse()` also creates a couple of clones to fill the `URL` struct). Since this is a private function behind a public options interface, I prefer to leave the deallocation `sentry__dsn_decref()`. This way, we have two clear points for deallocating all cloned strings: one in the options setter and one in `sentry_options_free()` and won't have to track for double-free, etc.

File: src/path/sentry_path_windows.c [hunk 5]
```diff
@@ -205,6 +220,25 @@ sentry__path_from_str(const char *s)
...
+    sentry_path_t *rv = SENTRY_MAKE(sentry_path_t);
+    if (!rv) {
+        return NULL;
+    }
+    rv->path = sentry_malloc(sizeof(wchar_t) * (s_len + 1));
+    if (!rv->path) {
+        sentry_free(rv);
+        return NULL;
+    }
+    MultiByteToWideChar(CP_ACP, 0, s, -1, rv->path, (int)s_len + 1);
```
[the remark this answers is not in the document]
Comment (supervacuus) [reply src/path/sentry_path_windows.c:238]: You are right; there are a couple of things wrong:

* We need explicit size (excluding 0) analog to explicit length
* The length of the converted string includes the null character (which only "works" if there was a zero immediately following the string slice)
* The destination string must be explicitly null-terminated (if we don't have a _correctly sized_ null-terminated input string because it will find 0 at some point).
* In case of an error in `MultiByteToWideChar`, we don't free `rv`

I also rewrote the null-terminated version in terms of the ptr/len version.

File: src/transports/sentry_transport_curl.c (hunk 2 above)
Comment (supervacuus) [reply src/transports/sentry_transport_curl.c:250]: I wanted to see if there was a trivial reason for `ASAN` failing to track `curl` globals when shutting down early (this is only related to the failing `ASAN` test on macOS and unrelated to the changes here, but it will fail again, so probably have to fix it here).

**Comment by supervacuus** [comment]
`LSAN` correctly found a leak, but it is outside our control: 

* our build on CI picks up the `libcurl` library of the `MacOSX 12.1 SDK` from `XCode 13.2.1`, which has version `7.77.0`, 
* the following fix was provided in [7.78.0](https://github.com/curl/curl/commit/62be0960858f18798f5f2bf662cbfd9ae3948eb4#diff-e0b647f3e21a7fb58400dcec74f381d2fe16ca41e983240c15fc1243c574dbf9) 
* which rectifies a leak introduced by a feature that was introduced with [7.77.0](https://github.com/curl/curl/commit/31f631a142d855f069242f3e0c643beec25d1b51#diff-e0b647f3e21a7fb58400dcec74f381d2fe16ca41e983240c15fc1243c574dbf9) (i.e., our version)

It is unclear why the build suddenly broke with the new runner images because neither XCode/SDK (and its `libcurl` version) nor llvm/clang/lsan where updated from previous builds.
